libevmasm/Ethdebug.cpp

```diff
@@ -0,0 +1,114 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libevmasm/Ethdebug.h>
+
+using namespace solidity;
+using namespace solidity::evmasm;
+using namespace solidity::evmasm::ethdebug;
+
+namespace
+{
+
+Json programInstructions(Assembly const* _assembly, LinkerObject const& _linkerObject, unsigned _sourceId)
+{
+	// e.g. interfaces don't have a valid assembly object.
+	if (_assembly)
+	{
+		solUnimplementedAssert(_assembly->eofVersion() == std::nullopt, "ethdebug does not yet support EOF.");
+		solUnimplementedAssert(_assembly->codeSections().size() == 1, "ethdebug does not yet support multiple code-sections.");
+	}
+
+	solAssert(_linkerObject.codeSectionLocations.size() == 1);
+	solAssert(_linkerObject.codeSectionLocations[0].end <= _linkerObject.bytecode.size());
+	Json instructions = Json::array();
+	for (size_t i = 0; i < _linkerObject.codeSectionLocations[0].instructionLocations.size(); ++i)
+	{
+		solAssert(_assembly);
+		LinkerObject::InstructionLocation currentInstruction = _linkerObject.codeSectionLocations[0].instructionLocations[i];
+		size_t start = currentInstruction.start;
+		size_t end = currentInstruction.end;
+		size_t assemblyItemIndex = currentInstruction.assemblyItemIndex;
+		solAssert(start < _linkerObject.bytecode.size());
+		solAssert(end <= _linkerObject.bytecode.size());
+		solAssert(start < end);
+		solAssert(assemblyItemIndex < _assembly->codeSections().at(0).items.size());
+		Json operation = Json::object();
+		operation["mnemonic"] = instructionInfo(static_cast<Instruction>(_linkerObject.bytecode[start]), _assembly->evmVersion()).name;
+		static size_t constexpr instructionSize = 1;
+		if (start + instructionSize < end)
+		{
+			bytes const argumentData(
+				_linkerObject.bytecode.begin() + static_cast<std::ptrdiff_t>(start) + instructionSize,
+				_linkerObject.bytecode.begin() + static_cast<std::ptrdiff_t>(end)
+			);
+			if (!argumentData.empty())
+				operation["arguments"] = Json::array({"0x" + util::toHex(argumentData)});
```

Comment:
Might just have used
```suggestion
				operation["arguments"] = Json::array({util::toHex(argumentData, HexPrefix::Add)});
```
since we have it (unless there's any reason not to use it here) - but also no reason to hold on merging.